### ISSUE TITLE
build 4.2.4 fails on PyString_Check with Python3

### ISSUE THREAD

**Issue by prjemian** [caqtdm/caqtdm#47]
Compiling with Python3, the build fails because it cannot find `PyString_Check`. 

Error dump:

```
mintadmin@mint-vm:~/.../eclipse/caqtdm$ make
cd caQtDM_QtControls/ && ( test -e Makefile || /usr/local/Qt/5.12.1/gcc_64/bin/qmake -o Makefile /home/mintadmin/Documents/eclipse/caqtdm/caQtDM_QtControls/caQtDM_QtControls.pro ) && make -f Makefile 
make[1]: Entering directory '/home/mintadmin/Documents/eclipse/caqtdm/caQtDM_QtControls'
make[1]: Nothing to be done for 'first'.
make[1]: Leaving directory '/home/mintadmin/Documents/eclipse/caqtdm/caQtDM_QtControls'
cd caQtDM_Lib/ && ( test -e Makefile || /usr/local/Qt/5.12.1/gcc_64/bin/qmake -o Makefile /home/mintadmin/Documents/eclipse/caqtdm/caQtDM_Lib/caQtDM_Lib.pro ) && make -f Makefile 
make[1]: Entering directory '/home/mintadmin/Documents/eclipse/caqtdm/caQtDM_Lib'
g++ -c -m64 -pipe -g -g -O2 -Wall -W -D_REENTRANT -fPIC -DDESIGNER_TOOLTIP_DESCRIPTIONS -DXDR_HACK -DXDR_LE -DIO_OPTIMIZED_FOR_TABWIDGETS -DTARGET_PRODUCT="\"Library for Display Manager\"" -DTARGET_FILENAME="\"caQtDM_Lib.dll\"" -DTARGET_VER_MAJ=4 -DTARGET_VER_MIN=2 -DTARGET_VER_BUILD=4_Development_a2f039d6 -DTARGET_COMPANY="\"Paul Scherrer Institut\"" -DTARGET_DESCRIPTION="\"Channel Access Qt Display Manager\"" -DTARGET_COPYRIGHT="\"Copyright (C) 2016 Paul Scherrer Institut\"" -DTARGET_INTERNALNAME="\"caqtdm\"" -DTARGET_VERSION_STR="\"V4.2.4_Development_a2f039d6\"" -DPYTHON -DQT_DISABLE_DEPRECATED_BEFORE=0x000000 -DQT_NO_DEBUG -DQT_PLUGIN -DQT_PRINTSUPPORT_LIB -DQT_UITOOLS_LIB -DQT_OPENGL_LIB -DQT_WIDGETS_LIB -DQT_GUI_LIB -DQT_NETWORK_LIB -DQT_CORE_LIB -I. -I/usr/local/epics/base-7.0.2/include/os/Linux -I/usr/local/epics/base-7.0.2/include/compiler/gcc -I/usr/include/python3.6m -I. -Isrc -IcaQtDM_Plugins -I../caQtDM_QtControls/src -I/usr/local/qwt-6.1.4/include -I/usr/local/epics/base-7.0.2/include -I/usr/local/Qt/5.12.1/gcc_64/include -I/usr/local/Qt/5.12.1/gcc_64/include/QtPrintSupport -I/usr/local/Qt/5.12.1/gcc_64/include/QtUiTools -I/usr/local/Qt/5.12.1/gcc_64/include/QtOpenGL -I/usr/local/Qt/5.12.1/gcc_64/include/QtWidgets -I/usr/local/Qt/5.12.1/gcc_64/include/QtGui -I/usr/local/Qt/5.12.1/gcc_64/include/QtNetwork -I/usr/local/Qt/5.12.1/gcc_64/include/QtCore -Imoc -isystem /usr/include/libdrm -I/usr/local/Qt/5.12.1/gcc_64/mkspecs/linux-g++-64 -o obj/caqtdm_lib.o src/caqtdm_lib.cpp
src/caqtdm_lib.cpp: In member function ‘void CaQtDM_Lib::HandleWidget(QWidget*, QString, bool, bool)’:
src/caqtdm_lib.cpp:1069:18: warning: variable ‘doit’ set but not used [-Wunused-but-set-variable]
             bool doit;
                  ^~~~
src/caqtdm_lib.cpp: In member function ‘bool CaQtDM_Lib::Python_Error(QWidget*, QString)’:
src/caqtdm_lib.cpp:3581:73: error: ‘PyString_Check’ was not declared in this scope
     if (errObj != NULL && (pystring = PyObject_Str(errObj)) != NULL && (PyString_Check(pystring))) {
                                                                         ^~~~~~~~~~~~~~
src/caqtdm_lib.cpp:3581:73: note: suggested alternative: ‘PyMapping_Check’
     if (errObj != NULL && (pystring = PyObject_Str(errObj)) != NULL && (PyString_Check(pystring))) {
                                                                         ^~~~~~~~~~~~~~
                                                                         PyMapping_Check
src/caqtdm_lib.cpp:3582:26: error: ‘PyString_AsString’ was not declared in this scope
        strcpy(errorType, PyString_AsString(pystring));
                          ^~~~~~~~~~~~~~~~~
src/caqtdm_lib.cpp:3582:26: note: suggested alternative: ‘PyBytes_AsString’
        strcpy(errorType, PyString_AsString(pystring));
                          ^~~~~~~~~~~~~~~~~
                          PyBytes_AsString
src/caqtdm_lib.cpp:3589:75: error: ‘PyString_Check’ was not declared in this scope
     if (errData != NULL && (pystring = PyObject_Str(errData)) != NULL && (PyString_Check(pystring))) {
                                                                           ^~~~~~~~~~~~~~
src/caqtdm_lib.cpp:3589:75: note: suggested alternative: ‘PyMapping_Check’
     if (errData != NULL && (pystring = PyObject_Str(errData)) != NULL && (PyString_Check(pystring))) {
                                                                           ^~~~~~~~~~~~~~
                                                                           PyMapping_Check
src/caqtdm_lib.cpp:3590:26: error: ‘PyString_AsString’ was not declared in this scope
        strcpy(errorInfo, PyString_AsString(pystring));
                          ^~~~~~~~~~~~~~~~~
src/caqtdm_lib.cpp:3590:26: note: suggested alternative: ‘PyBytes_AsString’
        strcpy(errorInfo, PyString_AsString(pystring));
                          ^~~~~~~~~~~~~~~~~
                          PyBytes_AsString
src/caqtdm_lib.cpp: In member function ‘void CaQtDM_Lib::DisplayContextMenu(QWidget*)’:
src/caqtdm_lib.cpp:6717:114: warning: format ‘%x’ expects argument of type ‘unsigned int’, but argument 5 has type ‘long int’ [-Wformat=]
                             snprintf(asc, MAX_STRING_LENGTH, "%ld (0x%x)", kPtr->edata.ivalue, kPtr->edata.ivalue);
                                                                                                ~~~~~~~~~~~~~~~~~~^
src/caqtdm_lib.cpp:6752:136: warning: format ‘%x’ expects argument of type ‘unsigned int’, but argument 5 has type ‘long int’ [-Wformat=]
                             snprintf(asc, MAX_STRING_LENGTH, "%ld (0x%x) %s", kPtr->edata.ivalue, kPtr->edata.ivalue, kPtr->edata.units);
                                                                                                   ~~~~~~~~~~~~~~~~~~                   ^
src/caqtdm_lib.cpp: In member function ‘virtual void CaQtDM_Lib::TreatRequestedValue(QString, QString, CaQtDM_Lib_Interface::FormatType, QWidget*)’:
src/caqtdm_lib.cpp:7630:9: warning: this statement may fall through [-Wimplicit-fallthrough=]
         }
         ^
src/caqtdm_lib.cpp:7633:5: note: here
     default:
     ^~~~~~~
Makefile:5145: recipe for target 'obj/caqtdm_lib.o' failed
make[1]: *** [obj/caqtdm_lib.o] Error 1
make[1]: Leaving directory '/home/mintadmin/Documents/eclipse/caqtdm/caQtDM_Lib'
Makefile:80: recipe for target 'sub-caQtDM_Lib-make_first' failed
make: *** [sub-caQtDM_Lib-make_first] Error 2
```

**Comment by prjemian** [comment]
According to this [SO post](https://stackoverflow.com/a/37105715), ``PyString_Check`` is deprecated.

**Comment by AntonMezger** [comment]
Hi Pete,

I made a modification to caQtDM_Lib.cpp and I join it to this email in order to try it out. I have actually no platform running with python3, so if you could me a favour and test it out for me, I would be happy.

Best regards

Anton

**Comment by prjemian** [comment]
Will do.  Thanks.

**Comment by prjemian** [comment]
Did not see that attachment.  Could send it again or post here as a new branch.

**Comment by AntonMezger** [comment]
rename the text file to .cpp
[caqtdm_lib.txt](https://github.com/caqtdm/caqtdm/files/2877383/caqtdm_lib.txt)

**Comment by prjemian** [comment]
still a couple errors on `PyString_AsString`, here are the last lines of the build console output:

```
g++ -c -m64 -pipe -g -g -O2 -Wall -W -D_REENTRANT -fPIC -DDESIGNER_TOOLTIP_DESCRIPTIONS -DXDR_HACK -DXDR_LE -DIO_OPTIMIZED_FOR_TABWIDGETS -DTARGET_PRODUCT="\"Library for Display Manager\"" -DTARGET_FILENAME="\"caQtDM_Lib.dll\"" -DTARGET_VER_MAJ=4 -DTARGET_VER_MIN=2 -DTARGET_VER_BUILD=4_Development_eec3b011 -DTARGET_COMPANY="\"Paul Scherrer Institut\"" -DTARGET_DESCRIPTION="\"Channel Access Qt Display Manager\"" -DTARGET_COPYRIGHT="\"Copyright (C) 2016 Paul Scherrer Institut\"" -DTARGET_INTERNALNAME="\"caqtdm\"" -DTARGET_VERSION_STR="\"V4.2.4_Development_eec3b011\"" -DPYTHON -DQT_DISABLE_DEPRECATED_BEFORE=0x000000 -DQT_NO_DEBUG -DQT_PLUGIN -DQT_PRINTSUPPORT_LIB -DQT_UITOOLS_LIB -DQT_OPENGL_LIB -DQT_WIDGETS_LIB -DQT_GUI_LIB -DQT_NETWORK_LIB -DQT_CORE_LIB -I. -I/usr/local/epics/base-7.0.2/include/os/Linux -I/usr/local/epics/base-7.0.2/include/compiler/gcc -I/usr/include/python3.6 -I. -Isrc -IcaQtDM_Plugins -I../caQtDM_QtControls/src -I/usr/local/qwt-6.1.4/include -I/usr/local/epics/base-7.0.2/include -I/usr/local/Qt/5.12.1/gcc_64/include -I/usr/local/Qt/5.12.1/gcc_64/include/QtPrintSupport -I/usr/local/Qt/5.12.1/gcc_64/include/QtUiTools -I/usr/local/Qt/5.12.1/gcc_64/include/QtOpenGL -I/usr/local/Qt/5.12.1/gcc_64/include/QtWidgets -I/usr/local/Qt/5.12.1/gcc_64/include/QtGui -I/usr/local/Qt/5.12.1/gcc_64/include/QtNetwork -I/usr/local/Qt/5.12.1/gcc_64/include/QtCore -Imoc -isystem /usr/include/libdrm -I/usr/local/Qt/5.12.1/gcc_64/mkspecs/linux-g++-64 -o obj/JSONValue.o src/JSONValue.cpp
src/caqtdm_lib.cpp: In member function ‘bool CaQtDM_Lib::Python_Error(QWidget*, QString)’:
src/caqtdm_lib.cpp:3587:26: error: ‘PyString_AsString’ was not declared in this scope
        strcpy(errorType, PyString_AsString(pystring));
                          ^~~~~~~~~~~~~~~~~
src/caqtdm_lib.cpp:3587:26: note: suggested alternative: ‘PyBytes_AsString’
        strcpy(errorType, PyString_AsString(pystring));
                          ^~~~~~~~~~~~~~~~~
                          PyBytes_AsString
src/caqtdm_lib.cpp:3600:26: error: ‘PyString_AsString’ was not declared in this scope
        strcpy(errorInfo, PyString_AsString(pystring));
                          ^~~~~~~~~~~~~~~~~
src/caqtdm_lib.cpp:3600:26: note: suggested alternative: ‘PyBytes_AsString’
        strcpy(errorInfo, PyString_AsString(pystring));
                          ^~~~~~~~~~~~~~~~~
                          PyBytes_AsString
src/caqtdm_lib.cpp: In member function ‘void CaQtDM_Lib::DisplayContextMenu(QWidget*)’:
src/caqtdm_lib.cpp:6725:114: warning: format ‘%x’ expects argument of type ‘unsigned int’, but argument 5 has type ‘long int’ [-Wformat=]
                             snprintf(asc, MAX_STRING_LENGTH, "%ld (0x%x)", kPtr->edata.ivalue, kPtr->edata.ivalue);
                                                                                                ~~~~~~~~~~~~~~~~~~^
src/caqtdm_lib.cpp:6760:136: warning: format ‘%x’ expects argument of type ‘unsigned int’, but argument 5 has type ‘long int’ [-Wformat=]
                             snprintf(asc, MAX_STRING_LENGTH, "%ld (0x%x) %s", kPtr->edata.ivalue, kPtr->edata.ivalue, kPtr->edata.units);
                                                                                                   ~~~~~~~~~~~~~~~~~~                   ^
g++ -c -m64 -pipe -g -g -O2 -Wall -W -D_REENTRANT -fPIC -DDESIGNER_TOOLTIP_DESCRIPTIONS -DXDR_HACK -DXDR_LE -DIO_OPTIMIZED_FOR_TABWIDGETS -DTARGET_PRODUCT="\"Library for Display Manager\"" -DTARGET_FILENAME="\"caQtDM_Lib.dll\"" -DTARGET_VER_MAJ=4 -DTARGET_VER_MIN=2 -DTARGET_VER_BUILD=4_Development_eec3b011 -DTARGET_COMPANY="\"Paul Scherrer Institut\"" -DTARGET_DESCRIPTION="\"Channel Access Qt Display Manager\"" -DTARGET_COPYRIGHT="\"Copyright (C) 2016 Paul Scherrer Institut\"" -DTARGET_INTERNALNAME="\"caqtdm\"" -DTARGET_VERSION_STR="\"V4.2.4_Development_eec3b011\"" -DPYTHON -DQT_DISABLE_DEPRECATED_BEFORE=0x000000 -DQT_NO_DEBUG -DQT_PLUGIN -DQT_PRINTSUPPORT_LIB -DQT_UITOOLS_LIB -DQT_OPENGL_LIB -DQT_WIDGETS_LIB -DQT_GUI_LIB -DQT_NETWORK_LIB -DQT_CORE_LIB -I. -I/usr/local/epics/base-7.0.2/include/os/Linux -I/usr/local/epics/base-7.0.2/include/compiler/gcc -I/usr/include/python3.6 -I. -Isrc -IcaQtDM_Plugins -I../caQtDM_QtControls/src -I/usr/local/qwt-6.1.4/include -I/usr/local/epics/base-7.0.2/include -I/usr/local/Qt/5.12.1/gcc_64/include -I/usr/local/Qt/5.12.1/gcc_64/include/QtPrintSupport -I/usr/local/Qt/5.12.1/gcc_64/include/QtUiTools -I/usr/local/Qt/5.12.1/gcc_64/include/QtOpenGL -I/usr/local/Qt/5.12.1/gcc_64/include/QtWidgets -I/usr/local/Qt/5.12.1/gcc_64/include/QtGui -I/usr/local/Qt/5.12.1/gcc_64/include/QtNetwork -I/usr/local/Qt/5.12.1/gcc_64/include/QtCore -Imoc -isystem /usr/include/libdrm -I/usr/local/Qt/5.12.1/gcc_64/mkspecs/linux-g++-64 -o obj/JSON.o src/JSON.cpp
g++ -m64 -pipe -g -g -O2 -Wall -W -D_REENTRANT -dM -E -o moc/moc_predefs.h /usr/local/Qt/5.12.1/gcc_64/mkspecs/features/data/dummy.cpp
/usr/local/Qt/5.12.1/gcc_64/bin/moc -DDESIGNER_TOOLTIP_DESCRIPTIONS -DXDR_HACK -DXDR_LE -DIO_OPTIMIZED_FOR_TABWIDGETS -DTARGET_PRODUCT="\"Library for Display Manager\"" -DTARGET_FILENAME="\"caQtDM_Lib.dll\"" -DTARGET_VER_MAJ=4 -DTARGET_VER_MIN=2 -DTARGET_VER_BUILD=4_Development_eec3b011 -DTARGET_COMPANY="\"Paul Scherrer Institut\"" -DTARGET_DESCRIPTION="\"Channel Access Qt Display Manager\"" -DTARGET_COPYRIGHT="\"Copyright (C) 2016 Paul Scherrer Institut\"" -DTARGET_INTERNALNAME="\"caqtdm\"" -DTARGET_VERSION_STR="\"V4.2.4_Development_eec3b011\"" -DPYTHON -DQT_DISABLE_DEPRECATED_BEFORE=0x000000 -DQT_NO_DEBUG -DQT_PLUGIN -DQT_PRINTSUPPORT_LIB -DQT_UITOOLS_LIB -DQT_OPENGL_LIB -DQT_WIDGETS_LIB -DQT_GUI_LIB -DQT_NETWORK_LIB -DQT_CORE_LIB --include /home/mintadmin/Documents/eclipse/caqtdm/caQtDM_Lib/moc/moc_predefs.h -I/usr/local/Qt/5.12.1/gcc_64/mkspecs/linux-g++-64 -I/home/mintadmin/Documents/eclipse/caqtdm/caQtDM_Lib -I'/home/mintadmin/Documents/eclipse/caqtdm/caQtDM_Lib//usr/local/epics/base-7.0.2/include/os/Linux' -I'/home/mintadmin/Documents/eclipse/caqtdm/caQtDM_Lib//usr/local/epics/base-7.0.2/include/compiler/gcc' -I'/home/mintadmin/Documents/eclipse/caqtdm/caQtDM_Lib//usr/include/python3.6' -I/home/mintadmin/Documents/eclipse/caqtdm/caQtDM_Lib -I/home/mintadmin/Documents/eclipse/caqtdm/caQtDM_Lib/src -I/home/mintadmin/Documents/eclipse/caqtdm/caQtDM_Lib/caQtDM_Plugins -I/home/mintadmin/Documents/eclipse/caqtdm/caQtDM_QtControls/src -I'/home/mintadmin/Documents/eclipse/caqtdm/caQtDM_Lib//usr/local/qwt-6.1.4/include' -I'/home/mintadmin/Documents/eclipse/caqtdm/caQtDM_Lib//usr/local/epics/base-7.0.2/include' -I/usr/local/Qt/5.12.1/gcc_64/include -I/usr/local/Qt/5.12.1/gcc_64/include/QtPrintSupport -I/usr/local/Qt/5.12.1/gcc_64/include/QtUiTools -I/usr/local/Qt/5.12.1/gcc_64/include/QtOpenGL -I/usr/local/Qt/5.12.1/gcc_64/include/QtWidgets -I/usr/local/Qt/5.12.1/gcc_64/include/QtGui -I/usr/local/Qt/5.12.1/gcc_64/include/QtNetwork -I/usr/local/Qt/5.12.1/gcc_64/include/QtCore -I/usr/include/c++/7 -I/usr/include/x86_64-linux-gnu/c++/7 -I/usr/include/c++/7/backward -I/usr/lib/gcc/x86_64-linux-gnu/7/include -I/usr/local/include -I/usr/lib/gcc/x86_64-linux-gnu/7/include-fixed -I/usr/include/x86_64-linux-gnu -I/usr/include src/caqtdm_lib.h -o moc/moc_caqtdm_lib.cpp
/usr/local/Qt/5.12.1/gcc_64/bin/moc -DDESIGNER_TOOLTIP_DESCRIPTIONS -DXDR_HACK -DXDR_LE -DIO_OPTIMIZED_FOR_TABWIDGETS -DTARGET_PRODUCT="\"Library for Display Manager\"" -DTARGET_FILENAME="\"caQtDM_Lib.dll\"" -DTARGET_VER_MAJ=4 -DTARGET_VER_MIN=2 -DTARGET_VER_BUILD=4_Development_eec3b011 -DTARGET_COMPANY="\"Paul Scherrer Institut\"" -DTARGET_DESCRIPTION="\"Channel Access Qt Display Manager\"" -DTARGET_COPYRIGHT="\"Copyright (C) 2016 Paul Scherrer Institut\"" -DTARGET_INTERNALNAME="\"caqtdm\"" -DTARGET_VERSION_STR="\"V4.2.4_Development_eec3b011\"" -DPYTHON -DQT_DISABLE_DEPRECATED_BEFORE=0x000000 -DQT_NO_DEBUG -DQT_PLUGIN -DQT_PRINTSUPPORT_LIB -DQT_UITOOLS_LIB -DQT_OPENGL_LIB -DQT_WIDGETS_LIB -DQT_GUI_LIB -DQT_NETWORK_LIB -DQT_CORE_LIB --include /home/mintadmin/Documents/eclipse/caqtdm/caQtDM_Lib/moc/moc_predefs.h -I/usr/local/Qt/5.12.1/gcc_64/mkspecs/linux-g++-64 -I/home/mintadmin/Documents/eclipse/caqtdm/caQtDM_Lib -I'/home/mintadmin/Documents/eclipse/caqtdm/caQtDM_Lib//usr/local/epics/base-7.0.2/include/os/Linux' -I'/home/mintadmin/Documents/eclipse/caqtdm/caQtDM_Lib//usr/local/epics/base-7.0.2/include/compiler/gcc' -I'/home/mintadmin/Documents/eclipse/caqtdm/caQtDM_Lib//usr/include/python3.6' -I/home/mintadmin/Documents/eclipse/caqtdm/caQtDM_Lib -I/home/mintadmin/Documents/eclipse/caqtdm/caQtDM_Lib/src -I/home/mintadmin/Documents/eclipse/caqtdm/caQtDM_Lib/caQtDM_Plugins -I/home/mintadmin/Documents/eclipse/caqtdm/caQtDM_QtControls/src -I'/home/mintadmin/Documents/eclipse/caqtdm/caQtDM_Lib//usr/local/qwt-6.1.4/include' -I'/home/mintadmin/Documents/eclipse/caqtdm/caQtDM_Lib//usr/local/epics/base-7.0.2/include' -I/usr/local/Qt/5.12.1/gcc_64/include -I/usr/local/Qt/5.12.1/gcc_64/include/QtPrintSupport -I/usr/local/Qt/5.12.1/gcc_64/include/QtUiTools -I/usr/local/Qt/5.12.1/gcc_64/include/QtOpenGL -I/usr/local/Qt/5.12.1/gcc_64/include/QtWidgets -I/usr/local/Qt/5.12.1/gcc_64/include/QtGui -I/usr/local/Qt/5.12.1/gcc_64/include/QtNetwork -I/usr/local/Qt/5.12.1/gcc_64/include/QtCore -I/usr/include/c++/7 -I/usr/include/x86_64-linux-gnu/c++/7 -I/usr/include/c++/7/backward -I/usr/lib/gcc/x86_64-linux-gnu/7/include -I/usr/local/include -I/usr/lib/gcc/x86_64-linux-gnu/7/include-fixed -I/usr/include/x86_64-linux-gnu -I/usr/include src/mutexKnobData.h -o moc/moc_mutexKnobData.cpp
/usr/local/Qt/5.12.1/gcc_64/bin/moc -DDESIGNER_TOOLTIP_DESCRIPTIONS -DXDR_HACK -DXDR_LE -DIO_OPTIMIZED_FOR_TABWIDGETS -DTARGET_PRODUCT="\"Library for Display Manager\"" -DTARGET_FILENAME="\"caQtDM_Lib.dll\"" -DTARGET_VER_MAJ=4 -DTARGET_VER_MIN=2 -DTARGET_VER_BUILD=4_Development_eec3b011 -DTARGET_COMPANY="\"Paul Scherrer Institut\"" -DTARGET_DESCRIPTION="\"Channel Access Qt Display Manager\"" -DTARGET_COPYRIGHT="\"Copyright (C) 2016 Paul Scherrer Institut\"" -DTARGET_INTERNALNAME="\"caqtdm\"" -DTARGET_VERSION_STR="\"V4.2.4_Development_eec3b011\"" -DPYTHON -DQT_DISABLE_DEPRECATED_BEFORE=0x000000 -DQT_NO_DEBUG -DQT_PLUGIN -DQT_PRINTSUPPORT_LIB -DQT_UITOOLS_LIB -DQT_OPENGL_LIB -DQT_WIDGETS_LIB -DQT_GUI_LIB -DQT_NETWORK_LIB -DQT_CORE_LIB --include /home/mintadmin/Documents/eclipse/caqtdm/caQtDM_Lib/moc/moc_predefs.h -I/usr/local/Qt/5.12.1/gcc_64/mkspecs/linux-g++-64 -I/home/mintadmin/Documents/eclipse/caqtdm/caQtDM_Lib -I'/home/mintadmin/Documents/eclipse/caqtdm/caQtDM_Lib//usr/local/epics/base-7.0.2/include/os/Linux' -I'/home/mintadmin/Documents/eclipse/caqtdm/caQtDM_Lib//usr/local/epics/base-7.0.2/include/compiler/gcc' -I'/home/mintadmin/Documents/eclipse/caqtdm/caQtDM_Lib//usr/include/python3.6' -I/home/mintadmin/Documents/eclipse/caqtdm/caQtDM_Lib -I/home/mintadmin/Documents/eclipse/caqtdm/caQtDM_Lib/src -I/home/mintadmin/Documents/eclipse/caqtdm/caQtDM_Lib/caQtDM_Plugins -I/home/mintadmin/Documents/eclipse/caqtdm/caQtDM_QtControls/src -I'/home/mintadmin/Documents/eclipse/caqtdm/caQtDM_Lib//usr/local/qwt-6.1.4/include' -I'/home/mintadmin/Documents/eclipse/caqtdm/caQtDM_Lib//usr/local/epics/base-7.0.2/include' -I/usr/local/Qt/5.12.1/gcc_64/include -I/usr/local/Qt/5.12.1/gcc_64/include/QtPrintSupport -I/usr/local/Qt/5.12.1/gcc_64/include/QtUiTools -I/usr/local/Qt/5.12.1/gcc_64/include/QtOpenGL -I/usr/local/Qt/5.12.1/gcc_64/include/QtWidgets -I/usr/local/Qt/5.12.1/gcc_64/include/QtGui -I/usr/local/Qt/5.12.1/gcc_64/include/QtNetwork -I/usr/local/Qt/5.12.1/gcc_64/include/QtCore -I/usr/include/c++/7 -I/usr/include/x86_64-linux-gnu/c++/7 -I/usr/include/c++/7/backward -I/usr/lib/gcc/x86_64-linux-gnu/7/include -I/usr/local/include -I/usr/lib/gcc/x86_64-linux-gnu/7/include-fixed -I/usr/include/x86_64-linux-gnu -I/usr/include src/MessageWindow.h -o moc/moc_MessageWindow.cpp
src/caqtdm_lib.cpp: In member function ‘virtual void CaQtDM_Lib::TreatRequestedValue(QString, QString, CaQtDM_Lib_Interface::FormatType, QWidget*)’:
src/caqtdm_lib.cpp:7638:9: warning: this statement may fall through [-Wimplicit-fallthrough=]
         }
         ^
src/caqtdm_lib.cpp:7641:5: note: here
     default:
     ^~~~~~~
/usr/local/Qt/5.12.1/gcc_64/bin/moc -DDESIGNER_TOOLTIP_DESCRIPTIONS -DXDR_HACK -DXDR_LE -DIO_OPTIMIZED_FOR_TABWIDGETS -DTARGET_PRODUCT="\"Library for Display Manager\"" -DTARGET_FILENAME="\"caQtDM_Lib.dll\"" -DTARGET_VER_MAJ=4 -DTARGET_VER_MIN=2 -DTARGET_VER_BUILD=4_Development_eec3b011 -DTARGET_COMPANY="\"Paul Scherrer Institut\"" -DTARGET_DESCRIPTION="\"Channel Access Qt Display Manager\"" -DTARGET_COPYRIGHT="\"Copyright (C) 2016 Paul Scherrer Institut\"" -DTARGET_INTERNALNAME="\"caqtdm\"" -DTARGET_VERSION_STR="\"V4.2.4_Development_eec3b011\"" -DPYTHON -DQT_DISABLE_DEPRECATED_BEFORE=0x000000 -DQT_NO_DEBUG -DQT_PLUGIN -DQT_PRINTSUPPORT_LIB -DQT_UITOOLS_LIB -DQT_OPENGL_LIB -DQT_WIDGETS_LIB -DQT_GUI_LIB -DQT_NETWORK_LIB -DQT_CORE_LIB --include /home/mintadmin/Documents/eclipse/caqtdm/caQtDM_Lib/moc/moc_predefs.h -I/usr/local/Qt/5.12.1/gcc_64/mkspecs/linux-g++-64 -I/home/mintadmin/Documents/eclipse/caqtdm/caQtDM_Lib -I'/home/mintadmin/Documents/eclipse/caqtdm/caQtDM_Lib//usr/local/epics/base-7.0.2/include/os/Linux' -I'/home/mintadmin/Documents/eclipse/caqtdm/caQtDM_Lib//usr/local/epics/base-7.0.2/include/compiler/gcc' -I'/home/mintadmin/Documents/eclipse/caqtdm/caQtDM_Lib//usr/include/python3.6' -I/home/mintadmin/Documents/eclipse/caqtdm/caQtDM_Lib -I/home/mintadmin/Documents/eclipse/caqtdm/caQtDM_Lib/src -I/home/mintadmin/Documents/eclipse/caqtdm/caQtDM_Lib/caQtDM_Plugins -I/home/mintadmin/Documents/eclipse/caqtdm/caQtDM_QtControls/src -I'/home/mintadmin/Documents/eclipse/caqtdm/caQtDM_Lib//usr/local/qwt-6.1.4/include' -I'/home/mintadmin/Documents/eclipse/caqtdm/caQtDM_Lib//usr/local/epics/base-7.0.2/include' -I/usr/local/Qt/5.12.1/gcc_64/include -I/usr/local/Qt/5.12.1/gcc_64/include/QtPrintSupport -I/usr/local/Qt/5.12.1/gcc_64/include/QtUiTools -I/usr/local/Qt/5.12.1/gcc_64/include/QtOpenGL -I/usr/local/Qt/5.12.1/gcc_64/include/QtWidgets -I/usr/local/Qt/5.12.1/gcc_64/include/QtGui -I/usr/local/Qt/5.12.1/gcc_64/include/QtNetwork -I/usr/local/Qt/5.12.1/gcc_64/include/QtCore -I/usr/include/c++/7 -I/usr/include/x86_64-linux-gnu/c++/7 -I/usr/include/c++/7/backward -I/usr/lib/gcc/x86_64-linux-gnu/7/include -I/usr/local/include -I/usr/lib/gcc/x86_64-linux-gnu/7/include-fixed -I/usr/include/x86_64-linux-gnu -I/usr/include src/limitsStripplotDialog.h -o moc/moc_limitsStripplotDialog.cpp
Makefile:5145: recipe for target 'obj/caqtdm_lib.o' failed
make[1]: *** [obj/caqtdm_lib.o] Error 1
make[1]: *** Waiting for unfinished jobs....
make[1]: Leaving directory '/home/mintadmin/Documents/eclipse/caqtdm/caQtDM_Lib'
Makefile:80: recipe for target 'sub-caQtDM_Lib-make_first' failed
make: *** [sub-caQtDM_Lib-make_first] Error 2
```

**Comment by prjemian** [comment]
My build os is linux mint 19.1 which provides Python 2.7 and 3.6.  Needed to load some additional libraries including the packages with .h files.  Installed Qt from their web site.  Here's my environment setup:

[setup_environment.sh.txt](https://github.com/caqtdm/caqtdm/files/2878409/setup_environment.sh.txt)

**Comment by AntonMezger** [comment]
one more try, if it does noes not work, I will have to install python3 myself.
[caqtdm_lib.txt](https://github.com/caqtdm/caqtdm/files/2879570/caqtdm_lib.txt)

**Comment by prjemian** [comment]
Great!  That built with no errors and the executable runs.  How can I test that the Python support is working as expected?

**Comment by prjemian** [comment]
CALC expressions: `PythonCalc`

**Comment by AntonMezger** [comment]
You may use tests.ui the cavalcade tab. You could then eventually also introduce an error in the python routine with the designer and verify you get a reasonable error message

Best regards

Anton

Envoyé de mon iPhone

**Comment by prjemian** [comment]
resolved in commit ba63473
